src/stats/archive/calculator.rs

```diff
@@ -0,0 +1,126 @@
+        Self { records }
+    }
+}
+
+impl ArchiveStats {
+    pub fn calculate_distribution<T>(
+        result: &BTreeMap<T, usize>,
+    ) -> Result<BTreeMap<usize, usize>, ArchiveCalculatorError> {
+        let mut distribution = BTreeMap::new();
+
+        for item in 0..*result
+            .values()
+            .max()
+            .ok_or(ArchiveCalculatorError::EmptyResult)?
+            + 1
+        {
+            distribution.insert(item, 0);
+        }
+
+        for count in result.values() {
+            let value = distribution.get_mut(count).unwrap();
+            *value += 1;
+        }
+
+        Ok(distribution)
+    }
+
+    pub fn number_of_tx_per_slot(&self) -> Result<BTreeMap<String, usize>, ArchiveCalculatorError> {
+        Ok(self
+            .records
+            .iter()
+            .group_by(|item| item.time.to_string())
+            .into_iter()
+            .map(|(key, group)| (key, group.count()))
+            .collect())
+    }
+
+    pub fn distinct_casters(&self) -> Result<BTreeSet<String>, ArchiveCalculatorError> {
+        Ok(self
+            .records
+            .iter()
+            .map(|item| item.caster.to_string())
+            .collect())
+    }
+
+    pub fn number_of_votes_per_caster(
+        &self,
+    ) -> Result<BTreeMap<String, usize>, ArchiveCalculatorError> {
+        Ok(self
+            .records
+            .iter()
+            .group_by(|item| item.caster.to_string())
+            .into_iter()
+            .map(|(key, group)| (key, group.count()))
+            .collect())
+    }
+
+    /// Method return max batch size calculated as biggest consequitve chain of transactions
+    /// sent to blockchain.
+    /// WARNING: single transaction in each slot would be counted as a batch also.
+    pub fn max_batch_size_per_caster(
+        &self,
+        slots_in_epoch: u32,
+    ) -> Result<BTreeMap<String, usize>, ArchiveCalculatorError> {
+        let time_era = self.records[0].time.time_era(slots_in_epoch);
+
+        Ok(self
+            .records
+            .iter()
+            .group_by(|item| item.caster.to_string())
+            .into_iter()
+            .map(|(key, group)| {
+                let mut sorted_group: Vec<&ArchiverRecord> = group.collect();
+
+                sorted_group.sort_by_key(|a| a.time);
+
+                let mut max_batch_size = 1;
+                let mut current_batch_size = 1;
+                let mut last_slot: BlockDate = sorted_group[0].time;
+
+                for item in sorted_group.iter().skip(1) {
+                    let current: BlockDate = item.time;
+                    if are_equal_or_adjacent(&last_slot, &current, &time_era) {
+                        current_batch_size += 1;
+                    } else {
+                        max_batch_size = std::cmp::max(max_batch_size, current_batch_size);
+                        current_batch_size = 1;
+                    }
+                    last_slot = current;
```

Comment:
this makes a few assumptions about the underlying blockchain / selection alg, for example that it follows a FIFO order